src/Arcus.EventGrid.Testing/Infrastructure/Hosts/ServiceBus/ServiceBusEventConsumerHost.cs

```diff
@@ -0,0 +1,144 @@
+        {
+            Guard.NotNullOrWhitespace(topicPath, nameof(topicPath));
+            Guard.NotNullOrWhitespace(serviceBusConnectionString, nameof(serviceBusConnectionString));
+            Guard.NotNull(logger, nameof(logger));
+
+            logger.LogInformation("Starting Service Bus event consumer host");
+
+            var managementClient = new ManagementClient(serviceBusConnectionString);
+            
+            var subscriptionName = $"Test-{Guid.NewGuid().ToString()}";
+            await CreateSubscriptionAsync(topicPath, managementClient, subscriptionName);
+            logger.LogInformation($"Created subscription '{subscriptionName}' on topic '{topicPath}'");
+            
+            var subscriptionClient = new SubscriptionClient(serviceBusConnectionString, topicPath, subscriptionName);
+            StartMessagePump(subscriptionClient, logger);
+            logger.LogInformation($"Message pump started on '{subscriptionName}' (topic '{topicPath}' for endpoint '{subscriptionClient.ServiceBusConnection?.Endpoint?.AbsoluteUri}')");
+
+            return new ServiceBusEventConsumerHost(topicPath, subscriptionName, subscriptionClient, managementClient, logger);
+        }
+
+        /// <summary>
+        ///     Stop receiving traffic
+        /// </summary>
+        public override async Task Stop()
+        {
+            _logger.LogInformation("Stopping host");
+            isHostShuttingDown = true;
+
+            await _managementClient.DeleteSubscriptionAsync(TopicPath, SubscriptionName);
+            _logger.LogInformation($"Subscription '{SubscriptionName}' deleted on topic '{TopicPath}'");
+
+            await _subscriptionClient.CloseAsync();
+
+            await base.Stop();
+        }
+
+        private static void StartMessagePump(SubscriptionClient subscriptionClient, ILogger logger)
+        {
+            var messageHandlerOptions = new MessageHandlerOptions(exceptionReceivedEventArgs => HandleException(exceptionReceivedEventArgs, logger))
+            {
+                AutoComplete = false,
+                MaxConcurrentCalls = 10
+            };
+
+            subscriptionClient.RegisterMessageHandler((receivedMessage, cancellationToken) => HandleNewMessage(receivedMessage, subscriptionClient, cancellationToken, logger), messageHandlerOptions);
+        }
+
+        private static async Task HandleNewMessage(Message receivedMessage, SubscriptionClient subscriptionClient, CancellationToken cancellationToken, ILogger logger)
+        {
+            if (receivedMessage == null || isHostShuttingDown)
+            {
+                return;
+            }
+
+            var rawReceivedEvents = Encoding.UTF8.GetString(receivedMessage.Body);
+
+            try
+            {
+                EventsReceived(rawReceivedEvents);
+
+                await subscriptionClient.CompleteAsync(receivedMessage.SystemProperties.LockToken);
+            }
+            catch (Exception)
+            {
```

Comment:
Would be good if we would keep the exception message in the log somehow, so that the real cause isn't lost.

Reply:
I changed it to:
```csharp
logger.LogError($"Failed to persist raw events with exception {ex.Message}. Payload: {rawReceivedEvents}");
```

Is this ok for you?

Reply:
Going to switch to structured logging as well